Adds group aggregation

thrasher-redhat opened the pull request:
Shiftzilla will now track data at the group level.

* Data is aggregated at a group level
* Each Group will now how its own breakdown page, similar to teams.
* The "Totals By Group" table will appear on the _overall page if Shiftzilla is built with the `--groups` flag
* Renames Test/Ops Blockers to the more general Blockers in several places
* Adds error to config check if a team name starts with "Group"
* Version bump to 0.2.27

thrasher-redhat committed:
Adds group aggregation

Shiftzilla will now track data at the group level.

* Data is aggregated at a group level
* Each Group will now how its own breakdown page, similar to teams.
* The "Totals By Group" table will appear on the _overall page if built with the `--groups` flag
* Renames Test/Ops Blockers to general blockers in several places
* Adds error to config check if a team name starts with "Group"
* Version bump to 0.2.27

thrasher-redhat commented:
@nhr Let me know if you want to meet and have me walk through this, as things are somewhat spreadout/confusing.

coverprice reviewed (approved):
idk ruby that well, but lgtm

nhr commented on lib/shiftzilla/config.rb:
```diff
@@ -22,6 +22,7 @@ def initialize
       cfg_file['Teams'].each do |team|
         @teams << Shiftzilla::Team.new(team,group_map)
       end
+      update_group_components
```
From a purity of essence perspective, I think I'd prefer `set_group_components`

nhr commented on lib/shiftzilla/group.rb:
```diff
@@ -1,10 +1,17 @@
     end
+
+    def set_components(component_list)
+      @components += component_list
```
This feels a little weird to me. "set_components" implies that we are do a 'write once' procedure, but the `+=` opens the door for being able to add more components later, and I don't think we want to do that.

thrasher-redhat replied:
I'll change it to use `||=`

nhr commented on lib/shiftzilla/org_data.rb:
```diff
@@ -200,8 +228,14 @@ def generate_reports
         }
 
+        # Check if this is actually a group "team"
+        if tname.start_with?("Group")
```
This makes me think we need something better than string matching to positively identify a group. Maybe we add a boolean `is_group?` method to the `tinfo` object? I know that in the config validator you've added a test to prevent teams with a name that starts with 'Group' but it doesn't feel like a robust solution.

thrasher-redhat replied:
I originally had that, but it only let me replace one or two of the string checks.  Its a bit weird, since we need to check for groups on the main page, but that `tinfo` is for `_overall`.  Its also a bit weird since one `populate_releases` ends up with a list of ordered team names - but not the objects themselves - so I ended up needing to check the group name using the starts_with to even set tinfo, which kind of negated the point.  I added a config check for teams that are named "Group*" to prevent issues.

I'll take a look tomorrow morning and see if I can add a is_group boolean to the teams.  I can't remember off the top of my head if I also need to be looking at the TeamData objects.

thrasher-redhat committed:
Replaces string checking to determine group status

Replaces the `start_with?("Group")` functions with checks for booleans that are
part of team_pinfo or team_files.

thrasher-redhat commented:
New commit to address comments.

* Replaced `update_group_components` with `set_group_components`
* Made `set_components` a one-time action
* Add `is_group` booleans to team_pinfo and team_files to avoid checking for "Group" in the name

nhr commented:
LGTM
